Index: tox.ini
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>[tox]\nenvlist = py{38,39,310,311,312,313,314}, lint\n\n[testenv]\nsetenv = PYTHONPATH = {toxinidir}/src\ndeps =\n    -r{toxinidir}/test_requirements.txt\ncommands =\n    pytest tests {posargs}\npassenv =\n    HOME\n    TMPDIR\n    USER\n    XDG_*\n\n[testenv:lint]\nsetenv = PYTHONPATH = {toxinidir}/src\nskip_install = True\ndeps =\n    black == 25.12.0\n    isort == 7.0.0\n    flake8 == 7.3.0\n;    mypy == 1.19.0\ncommands =\n    black --check {toxinidir}/src {toxinidir}/tests\n    isort --check {toxinidir}/src {toxinidir}/tests\n    flake8 {toxinidir}/src {toxinidir}/tests\n;    mypy {toxinidir}/src\n\n[gh-actions]\npython =\n    3.8: py38\n    3.9: py39\n    3.10: py310\n    3.11: py311\n    3.12: py312\n    3.13: py313\n    3.14: py314\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tox.ini b/tox.ini
--- a/tox.ini	(revision 167411d251e815679e64fe2b927f94da5339e8cf)
+++ b/tox.ini	(date 1765704387336)
@@ -20,12 +20,12 @@
     black == 25.12.0
     isort == 7.0.0
     flake8 == 7.3.0
-;    mypy == 1.19.0
+    mypy == 1.19.0
 commands =
     black --check {toxinidir}/src {toxinidir}/tests
     isort --check {toxinidir}/src {toxinidir}/tests
     flake8 {toxinidir}/src {toxinidir}/tests
-;    mypy {toxinidir}/src
+    mypy {toxinidir}/src
 
 [gh-actions]
 python =
